include/nbl/video/IDescriptorPool.h

```diff
@@ -32,86 +30,186 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
-            m_mutableSamplerStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<video::IGPUSampler>>[]>(m_maxDescriptorCount[asset::EDT_COMBINED_IMAGE_SAMPLER]);
-            m_storageImageStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
-            m_UBO_SSBOStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_UTB_STBStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_accelerationStructureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
-        }
-
-        virtual ~IDescriptorPool() {}
-
-        virtual core::smart_refctd_ptr<IGPUDescriptorSet> createDescriptorSet_impl(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout) = 0;
-        virtual bool freeDescriptorSets_impl(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets) = 0;
+            inline void reset(const bool allowsFreeing)
+            {
+                if (!allowsFreeing)
+                    linearAllocator.reset();
+                else
+                    generalAllocator.reset();
+            }
 
-        uint32_t m_maxSets;
+            inline uint32_t getAllocatedDescriptorCount(const bool allowsFreeing) const
+            {
+                if (!allowsFreeing)
+                    return linearAllocator.get_allocated_size();
+                else
+                    return generalAllocator.get_allocated_size();
+            }
 
-    private:
-        const IDescriptorPool::E_CREATE_FLAGS m_flags;
-        uint32_t m_maxDescriptorCount[asset::EDT_COUNT];
-        union
-        {
-            core::LinearAddressAllocator<uint32_t> m_linearAllocators[asset::EDT_COUNT+1];
-            core::GeneralpurposeAddressAllocator<uint32_t> m_generalAllocators[asset::EDT_COUNT+1];
+            union
+            {
+                core::LinearAddressAllocator<uint32_t> linearAllocator;
+                core::GeneralpurposeAddressAllocator<uint32_t> generalAllocator;
+            };
+            std::unique_ptr<uint8_t[]> generalAllocatorReservedSpace = nullptr;
         };
-        std::unique_ptr<uint8_t[]> m_generalAllocatorReservedSpace[asset::EDT_COUNT+1];
+        std::unique_ptr<allocator_state_t> m_descriptorAllocators[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT) + 1];
```

Comment:
why unique ptr, they can be regular members ?

Reply:
Can't be, one of the two allocators doesn't have a default ctor which makes `allocator_state_t` not have a default ctor, and I need `m_creationParameters` to create  `allocator_state_t` so they are `std::unique_ptr` to defer their creation in the ctor body.

Reply:
ok I guess, it would be vastly more code to add a move ctor and default ctor